horovod/common/operations.cc

```diff
@@ -200,17 +200,28 @@ OperationManager* CreateOperationManager(HorovodGlobalState& state) {
   }
 #endif
 
+  std::shared_ptr<JoinOp> join_op(new JoinOp(&state));
   std::shared_ptr<ErrorOp> error_op(new ErrorOp(&state));
 
   return new OperationManager(&state.parameter_manager, allreduce_ops,
-                              allgather_ops, broadcast_ops, error_op);
+                              allgather_ops, broadcast_ops, join_op, error_op);
 }
 
 // Process a Response by doing a reduction, a gather, a broadcast, or
 // raising an error.
-void PerformOperation(Response response) {
+void PerformOperation(Response response, HorovodGlobalState& state) {
+  if (response.response_type() == Response::JOIN) {
```

Comment:
Making this into a special case seems strange since you already have the `JoinOp`, which currently does nothing (throws `PreconditionError`).  Can we just move this logic into the `JoinOp` to make the code more general?